### ISSUE TITLE
Perform Fast Fourier Transform in C++

### ISSUE THREAD

**Issue by yijiang1** [OpenChemistry/tomviz#451]
Is there a way to do FFT in C++ now?
If yes, can some one provide me a sample c++ script as a quick tutorial? Thanks.

**Comment by Hovden** [comment]
@cquammen @cryos Can we readily call the FFT from VTK (http://www.vtk.org/doc/nightly/html/classvtkImageFFT.html) or ITK (https://itk.org/Doxygen/html/group__ITKFFT.html)?

I assume we have those bundled into tomviz. If so, a short sample code of their implementation can get Yi started on advanced reconstruction algorithms.

**Comment by cryos** [comment]
Sorry about the delay in replying, @mathturtle do you have any time to look into this? I haven't used either class, but we should be able to get an example together. I would suggest the ITK class as it abstracts FFTW nicely. We would have to link to an ITK library, so it will require some modification of the build system too (we are already finding ITK now).

**Comment by mathturtle** [comment]
So I don't know much about the ITK one but I can take a look at it.  I think from discussions with @cquammen, the issue we foresee is that it only takes itkImages as its input.  So we will have to convert the data into itkImage and back, which will probably require the ITKVTK bridge.  It may be better to just link to fftw directly and use that.

**Comment by Hovden** [comment]
It does not necessarily have to be FFTW, a simple FFT function might work well enough.

**Comment by yijiang1** [comment]
Yeah, I just want to do FFT in C++ right now. I don't really care which library I use.

**Comment by cryos** [comment]
OK, @mathturtle can you just take a look at getting a C++ example together for the VTK class, and we can work with them to confirm it has the basic features needed. The tests are Python and Tcl, perhaps convert one of them to C++, verify it, and make a pull request.

**Comment by mathturtle** [comment]
So I have been reading over the APIs for FFTW, ITKFFT, and VTK's FFT.  It looks like the most appropriate to use will depend on the data type that you have.  If you have an ITK image, you'll want to use the ITK one.  If you have a VTK image then the VTK one will be the easiest to use.  And if you have a raw array of floats then it will probably be easiest to use FFTW directly (this may also be the best for the VTK image case, I'm not sure).  I will start with the VTK example, but depending on what we want to use, I can probably put together a FFTW example too.

**Comment by cryos** [comment]
VTK is great for now, let's not overthink it too much until it starts being used - vtkImageData is going to be the main target.
